In [1]:
from pathlib import Path
import pandas as pd

PROJECT_ROOT = Path.cwd().parent

FULL_INFO_FILE = (
    PROJECT_ROOT
    / "data"
    / "processed"
    / "student_full_information.csv"
)

X_full = pd.read_csv(FULL_INFO_FILE)

print("X_full shape:", X_full.shape)
X_full.head()

X_full shape: (395, 41)


,age,Medu,Fedu,traveltime,studytime,failures,famrel,freetime,goout,Dalc,...,guardian_mother,guardian_other,schoolsup_yes,famsup_yes,paid_yes,activities_yes,nursery_yes,higher_yes,internet_yes,romantic_yes
0,18,4,4,2,2,0,4,3,4,1,...,True,False,True,False,False,False,True,True,False,False
1,17,1,1,1,2,0,5,3,3,1,...,False,False,False,True,False,False,False,True,True,False
2,15,1,1,1,2,3,4,3,2,2,...,True,False,True,False,True,False,True,True,True,False
3,15,4,2,1,3,0,3,2,2,1,...,True,False,False,True,True,True,True,True,True,True
4,16,3,3,1,2,0,4,3,2,1,...,False,False,False,True,True,False,True,True,False,False


In [2]:
leak_cols = [
    c for c in X_full.columns
    if c in ("G1", "G2")
]

X_early = X_full.drop(columns=leak_cols)

print("Early-warning features:", X_early.shape[1])

Early-warning features: 39


In [3]:
print("G1 present:", "G1" in X_early.columns)
print("G2 present:", "G2" in X_early.columns)

G1 present: False
G2 present: False


In [4]:
print("Full-information shape:", X_full.shape)
print("Early-warning shape:", X_early.shape)

Full-information shape: (395, 41)
Early-warning shape: (395, 39)


In [5]:
assert "G1" not in X_early.columns
assert "G2" not in X_early.columns

print("Validation passed.")

Validation passed.


In [6]:
OUTPUT_FILE = (
    PROJECT_ROOT
    / "data"
    / "processed"
    / "student_early_warning.csv"
)

X_early.to_csv(
    OUTPUT_FILE,
    index=False
)

print("Saved:", OUTPUT_FILE)

Saved: /Users/jjun28/Desktop/Github 2026/student-performance-prediction-ml/data/processed/student_early_warning.csv


# Session 21: Early-Warning Dataset Design

## Objective

Create an early-warning feature set by removing G1 and G2 from the full-information dataset.

## Results

- Full-information shape: (395, 41)
- Early-warning shape: (395, 39)

## Interpretation

The early-warning dataset excludes prior grades G1 and G2. These variables are highly predictive of final academic performance but may not be available early enough in the academic term for intervention purposes.

Removing G1 and G2 creates a more realistic early-warning scenario where predictions are made using demographic, family, behavioral, and school-related information available before major grades are recorded.